### PR TITLE
CreateAnd*Patch now GETs the object if it already exists

### PR DESCRIPTION
**Pull request by rfranzke**
**How to categorize this PR?**

/area dev-productivity
/kind task

**What this PR does / why we need it**:
As discussed in https://github.com/gardener/gardener/pull/4027#issuecomment-851942595 and ff.

/invite @timebertt 

**Release note**:

```other operator
NONE
```

### CODE REVIEW

File: pkg/controllerutils/patch.go [hunk 1]
```diff
@@ -74,33 +70,25 @@ func getAndCreateOrPatch(ctx context.Context, c client.Client, obj client.Object
...
 //
 // It returns the executed operation and an error.
-func CreateOrStrategicMergePatch(ctx context.Context, c client.Writer, obj client.Object, f controllerutil.MutateFn) (controllerutil.OperationResult, error) {
-	return createOrPatch(ctx, c, obj, func(obj client.Object) client.Patch { return client.StrategicMergeFrom(obj) }, f)
+func CreateOrGetAndStrategicMergePatch(ctx context.Context, c client.Client, obj client.Object, f controllerutil.MutateFn) (controllerutil.OperationResult, error) {
+	return createOrGetAndPatch(ctx, c, obj, func(obj client.Object) client.Patch { return client.StrategicMergeFrom(obj) }, f)
 }
 
-func createOrPatch(ctx context.Context, c client.Writer, obj client.Object, patchFunc func(client.Object) client.Patch, f controllerutil.MutateFn) (controllerutil.OperationResult, error) {
+func createOrGetAndPatch(ctx context.Context, c client.Client, obj client.Object, patchFunc func(client.Object) client.Patch, f controllerutil.MutateFn) (controllerutil.OperationResult, error) {
```
Comment (timebertt) [inline pkg/controllerutils/patch.go:93]: As discussed, we need to 
- calculate the patch based on the object after we read it from the API server / cache (i.e. after `Get` but before `f()`)
- reset the object to an empty object (obtain it from `c.Scheme().New`) when we received an `AlreadyExistsError` to unset optional fields that were set in `f()`. If we don't unset optional fields added in `f()`, `Get` will not unset them for us and thus we won't set them via the patch. We need something like this: https://github.com/gardener/gardener/blob/893f4042fa3e4a1814ab04b4dd79e9a09e1f914a/pkg/extensions/customresources.go#L520-L543

**Review by timebertt — CHANGES_REQUESTED** [review]

File: pkg/controllerutils/patch.go (hunk 1 above)
Comment (rfranzke) [reply pkg/controllerutils/patch.go:93]: Thanks, done!

File: pkg/controllerutils/patch_test.go [hunk 2]
```diff
@@ -185,6 +208,10 @@ var _ = Describe("Patch", func() {
 
 				gomock.InOrder(
 					c.EXPECT().Create(ctx, obj).Return(apierrors.NewAlreadyExists(schema.GroupResource{}, "")),
+					c.EXPECT().Get(ctx, client.ObjectKeyFromObject(obj), obj).DoAndReturn(func(_ context.Context, _ client.ObjectKey, objToReturn *corev1.ServiceAccount) error {
+						obj.DeepCopyInto(objToReturn)
+						return nil
```
Comment (timebertt) [inline pkg/controllerutils/patch_test.go:215]: Can you add the following to assert that the reset logic is working properly?

```suggestion
						obj.DeepCopyInto(objToReturn)
												// mimic behavior of json decoder: if optional fields are not set on the server, it won't unset them on the in-memory object.
						// In this case: the mutate func was already called for the create call, so the in-memory object will have the labels, but the
						// get call won't remove them.
						objToReturn.Labels = objCopy.Labels
						return nil
```

With this, the test will fail, if the reset func is not called. Currently, all tests will go green, when you remove the reset call.

**Review by timebertt — COMMENTED** [review]
Very nice, thank you!
Only one suggestion for improving the test :)

File: pkg/controllerutils/patch_test.go (hunk 2 above)
Comment (timebertt) [reply pkg/controllerutils/patch_test.go:215]: I wanted to try out the "allow edits by maintainers" feature and pushed the change as another commit myself, hope you don't mind 😉 
Let me know what you think :)

Comment (rfranzke) [reply pkg/controllerutils/patch_test.go:215]: Why are the tests failing now after your change?

Comment (timebertt) [reply pkg/controllerutils/patch_test.go:215]: Well, I figured, my change didn't make any sense 🙈 
I pushed another commit that asserts that our reset logic is working, so what I originally wanted to do :)

Comment (rfranzke) [reply pkg/controllerutils/patch_test.go:215]: OK, this looks good. Who approves this PR now? 😉

Comment (timebertt) [reply pkg/controllerutils/patch_test.go:215]: I guess, we both? 🤔😁

**Review by timebertt — APPROVED** [review]
/lgtm

File: pkg/controllerutils/patch_test.go (hunk 2 above)
Comment (rfranzke) [reply pkg/controllerutils/patch_test.go:215]: image

approve :)

**Pull request merged**